### PR TITLE
feat(platform/bitbucket-server): allow blobless clone

### LINKED ISSUE #27853: Use blobless git with Bitbucket Server / DataCenter
### Discussed in https://github.com/renovatebot/renovate/discussions/27849

<div type='discussions-op-text'>

<sup>Originally posted by **phaldan** March 11, 2024</sup>
### Tell us more.

Since #11759 a full clone is used for Atlassian Bitbucket due to missing support for blobless git. As mentioned in the issue this should be supported by Bitbucket since 7.7. Only Bitbucket 8 is currently still supported and it should be safe to use blobless git with Bitbucket:
https://confluence.atlassian.com/support/atlassian-support-end-of-life-policy-201851003.html</div>

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Fetch `bbServer` version from api and use it to decide cloning method

## Context
Closes: https://github.com/renovatebot/renovate/issues/27853

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (tested locally)

### CODE REVIEW

File: lib/modules/platform/bitbucket-server/index.ts [hunk 1]
```diff
@@ -98,6 +101,29 @@ export function initPlatform({
   const platformConfig: PlatformResult = {
     endpoint: defaults.endpoint,
   };
+  let bitbucketServerVersion: string;
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:104]: move to try catch

File: lib/modules/platform/bitbucket-server/index.ts [hunk 2]
```diff
@@ -98,6 +101,29 @@ export function initPlatform({
...
+    }
+
+    logger.debug('Bitbucket Server version is: ' + bitbucketServerVersion);
+    defaults.version = bitbucketServerVersion;
+  } catch (err) {
+    logger.debug(
+      { err },
+      'Error authenticating with Bitbucket. Check that your token includes "api" permissions',
+    );
+    throw new Error('Init: Authentication failure');
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:124]: I don't think we should throw here

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 3]
```diff
@@ -247,6 +251,10 @@ describe('modules/platform/bitbucket-server/index', () => {
         });
 
         it('should init', async () => {
+          httpMock
+            .scope(urlHost)
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:255]: wrong url

Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.spec.ts:255]: Sorry, I am not sure how this is wrong? I logged them and they seemed correct, also the other tests use the same.

File: lib/modules/platform/bitbucket-server/index.ts [hunk 4]
```diff
@@ -98,6 +101,28 @@ export function initPlatform({
   const platformConfig: PlatformResult = {
     endpoint: defaults.endpoint,
   };
+  try {
+    let bitbucketServerVersion: string;
+    if (process.env.RENOVATE_X_PLATFORM_VERSION) {
```
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/index.ts:106]: Why does this need to be added?

File: lib/modules/platform/bitbucket-server/index.ts [hunk 5]
```diff
@@ -98,6 +101,28 @@ export function initPlatform({
...
+      bitbucketServerVersion = process.env.RENOVATE_X_PLATFORM_VERSION;
+    } else {
+      const { version } = (
+        await bitbucketServerHttp.getJson<{ version: string }>(
+          `./rest/api/1.0/application-properties`,
+        )
+      ).body;
+      bitbucketServerVersion = version;
```
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/index.ts:114]: Check that it's valid semver

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket-server/index.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.ts:106]: if the user provides this value, we can save on api call

File: lib/modules/platform/bitbucket-server/index.ts [hunk 6]
```diff
@@ -98,7 +101,32 @@ export function initPlatform({
...
+      logger.debug('Bitbucket Server version is: ' + bitbucketServerVersion);
+    }
+
+    if (semver.valid(bitbucketServerVersion)) {
```
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/index.ts:119]: Does it return false for undefined?

Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket-server/index.ts:119]: String / Null

**Review by rarkins — DISMISSED** [review]

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 7]
```diff
@@ -211,6 +213,7 @@ describe('modules/platform/bitbucket-server/index', () => {
         // reset module
         jest.resetModules();
         bitbucket = await import('.');
+        logger = (await import('../../../logger')).logger as any;
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:216]:
```suggestion
        logger = mocked(await import('../../../logger')).logger;
```

File: lib/modules/platform/bitbucket-server/index.ts [hunk 8]
```diff
@@ -98,7 +101,32 @@ export function initPlatform({
   const platformConfig: PlatformResult = {
     endpoint: defaults.endpoint,
   };
-  return Promise.resolve(platformConfig);
+  try {
+    let bitbucketServerVersion: string;
+    // istanbul ignore if: experimental feature
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:106]: nope

File: lib/modules/platform/bitbucket-server/index.ts [hunk 9]
```diff
@@ -98,7 +101,32 @@ export function initPlatform({
...
+    if (process.env.RENOVATE_X_PLATFORM_VERSION) {
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:107]: Add to experimental docs

File: lib/modules/platform/bitbucket-server/index.ts (hunk 4 above)
Comment (viceice) [reply lib/modules/platform/bitbucket-server/index.ts:107]: @rarkins We've same for gitlab.

**Review by viceice — APPROVED** [review]

**Pull request merged**